lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -380,7 +380,12 @@ public function commit($entity = null)
         try {
             // Collection deletions (deletions of complete collections)
             foreach ($this->collectionDeletions as $collectionToDelete) {
-                $this->getCollectionPersister($collectionToDelete->getMapping())->delete($collectionToDelete);
+                if (!$collectionToDelete instanceof PersistentCollection ||
+                    // deferred explicit tracked collections can be removed only when owning relation was persisted
+                    $this->em->getClassMetadata(get_class($owner = $collectionToDelete->getOwner()))->isChangeTrackingDeferredImplicit() || 
+                    $this->isScheduledForDirtyCheck($owner)) {
+                    $this->getCollectionPersister($collectionToDelete->getMapping())->delete($collectionToDelete);
+                }
```

Comment:
I think there's a bit too much happening here. We can split things, do an early exit for non-persistent collections, and remove the assignment in condition.

Reply:
We can't do early exit, statement must be executed for non-persistent collections. And assignment needs to happen for persistent collections only.

Reply:
`continue` is an early-exit for the loop :)
I just pushed the changes :+1:

Reply:
Yeah I know. I meant we can't do it without duplicating code. If that's fine, then ok